### ISSUE TITLE
Profiler doesnt sum to total calculation time

### ISSUE THREAD

**Issue by erwanp** [radis/radis#322]
### 🐛 Describe the bug

Total time registered in the profiler doesn't sum to the total calculation time. Actually, it's about 2x the calculation time, because some parameters are counted twice, for instance `part_function` and `calc_noneq_population_multiTvib` 

### 🖥 Steps To Reproduce

```python
from radis import calc_spectrum
s  = calc_spectrum(2284.2, 2284.6,
                     Trot=491,
                     Tvib=(517, 517, 2641),
                     wstep='auto', 
                     pressure=20*1e-3,           # bar
                     molecule='CO2',
                     isotope='1,2',
                     path_length=10,             # cm-1
                     mole_fraction=0.1*28.97/44.07,
                     broadening_max_width=1,     # cm-1
                     medium='vacuum',
                     databank='HITEMP-CO2-TEST',
                     )

print(s.conditions['profiler'])
>>> {'fetch_energy_6': 0.9933407306671143, 'calc_weight_trans': 0.08555841445922852, 'check_non_eq_param': 1.1167986392974854, 'part_function': 0.13460326194763184, 'populations': 0.00997471809387207, 'calc_noneq_population_multiTvib': 0.1445779800415039, 'map_part_func': 0.0, 'corrected_population_se': 0.003988027572631836, 'scaled_non_eq_linestrength': 0.003988027572631836, 'calc_emission_integral': 0.01595783233642578, 'applied_linestrength_cutoff': 0.003989458084106445, 'calc_lineshift': 0.0009968280792236328, 'calc_hwhm': 0.014960289001464844, 'precompute_DLM_lineshapes': 0.0009970664978027344, 'DLM_Initialized_vectors': 0.0, 'DLM_closest_matching_line': 0.0, 'DLM_Distribute_lines': 0.0, 'DLM_convolve': 0.0, 'calc_line_broadening': 0.0019965171813964844, 'calc_other_spectral_quan': 0.0049893856048583984, 'spectrum_calc_before_obj': 1.3212203979492188}

print(sum(list(s.conditions["profiler"].values())))
>>> 3.8579375743865967

print(s.conditions['calculation_time'])
>>> 1.3212203979492188
```

### 🎯 Expected behavior

Store calculation times should be a dict with nested levels. 
Total of each level should be (approximately) correct
We should add a test (with some tolerance)

**Comment by anandxkumar** [comment]
We can have a format like below for `s.conditions['profiler']`:
```
{
1: {'calc_xyz': 2.5423}
2: {'calc_1': 2.5, 'calc_2': 0.11,....}
3:{'calc_3': 2.5, 'calc_4': 0.11,....}
}
```
Now the issues that we are facing:
eg:
```
sf = SpectrumFactory(wavenum_min=2900,
                     wavenum_max=4200,
                     molecule='CO',
                     broadening_max_width=10,  # cm-1
                     medium='vacuum',
                     verbose=0,    # more for more details
                     pressure = 10,
                     wstep = 0.01,
                     )
sf.fetch_databank('hitemp')
s1 = sf.eq_spectrum(Tgas=300, path_length=1)

new_dict = {1:{},2:{},3:{}}
dict_org = sf.profiler.dict_time
for i in dict_org:
    new_dict[dict_org[i][1]].update({i:dict_org[i][0]})

print(new_dict)
```
We get the output as the below:

```
{1: {'spectrum_calculation': 1.8847501277923584},
 2: {'scaled_eq_linestrength': 0.011523962020874023,
  'calc_lineshift': 0.0011229515075683594,
  'calc_hwhm': 0.009284019470214844,
  'calc_line_broadening': 1.8466684818267822,
  'calc_other_spectral_quan': 0.009803056716918945,
  'spectrum_calc_before_obj': 1.8839342594146729,
  'generate_spectrum_obj': 0.0008108615875244141},
 3: {'precompute_DLM_lineshapes': 0.05222773551940918,
  'DLM_Initialized_vectors': 5.245208740234375e-06,
  'DLM_closest_matching_line': 0.004378080368041992,
  'DLM_Distribute_lines': 0.022677183151245117,
  'DLM_convolve': 1.7614152431488037}}
```

- If we see level 2, there is 'spectrum_calc_before_obj' which adds all the previous step and is present here, also is an important element for analysis, so the sum will be almost 2x in verbose 2. 
- Not all piece of code has been included in Profiler. As of now, only `Factory.py,` `base.py`, and `broadening.py` have been converted. `loader.py` gives an error due to the issue that the profiler isn't declared till the spectrum is calculated eg eq_spectrum, non_eq_spectrum, etc. Also, some pieces of code are covered in verbose level 2 but not level 3, for eg- `_calc_broadening_HWHM` uses level 2 for all calculations [link](https://github.com/radis/radis/blob/9bb5d7b7fb153dfed9116fa051919446bdd38508/radis/lbl/broadening.py#L815), so definitely some difference will arise due to these. Fix will be to re-write the level accordingly, and there might be places that don't require level 3. So tolerance will need to be increased.
- Due to the limited code been covered, the calculation time at different levels can be significantly different (more if the total calculation time is less, so high tolerance is required there in that case).

**Comment by erwanp** [comment]
What you would really need is a tree structure. It's easy to build one in Python, but it should be able to be serialized/deserialized when the Spectrum object is stored as a .spec (JSON file). [This](https://stackoverflow.com/a/23597335/5622825) may be helpful. You would need to add Tests to make sure it's serialized/deserialized properly. 

A more straightforward solution would be to use our own convention with a dict class: 
```
Node = {"value": float , "subtasks": Node} 
```
e.g : 
```
profiler = OrderedDict{
    "step1": {"value":0.1 }  ,
    "step2": # with subtasks : {"value": 0.8,  "subtasks" :{"step2a":"0.2", "step2b":0.45}},
    "step3": 0.3   # if no subtasks we could skip the "value"
    }
```
Sounds like a simpler/better solution : you won't have to worry about serializing/deserializing the data. You can then update the display functions based on this structure. 

OrderedDict may be better suited than dict to keep the calculation states ordered. Note :  I remember OrderedDict is already serialized/deserialized properly by [json-tricks](https://pypi.org/project/json-tricks/), which we use when storing .spec files. 


---

About your notes : 
- if `spectrum_calc_before_obj` is level2 but sums other durations from level2, then it should be removed. 
- that not all piece of codes are included in Profiler is fine. We should just not count something twice. If the difference between total time and subtasks is significant then we need to cover the missing subtasks. 
- `generate_spectrum_obj` significance was an error fixed by #304
